propagation/amazon.go

```diff
@@ -0,0 +1,83 @@
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	amazonTracePropagationHTTPHeader = "X-Amzn-Trace-Id"
+)
+
+// MarshalAmazonTraceContext uses the information in prop to create a trace context header
+// in the Amazon X-Ray trace header format. It returns the serialized form of the trace
+// context, ready to be inserted into the headers of an outbound HTTP request.
+func MarshalAmazonTraceContext(prop *PropagationContext) string {
+	if prop == nil {
+		return ""
+	}
+
+	if prop.TraceID == "" || prop.ParentID == "" {
+		return ""
+	}
+
+	h := fmt.Sprintf("Root=%s;Parent=%s", prop.TraceID, prop.ParentID)
+
+	if len(prop.TraceContext) != 0 {
+		elems := make([]string, len(prop.TraceContext))
+		i := 0
+		for k, v := range prop.TraceContext {
+			elems[i] = fmt.Sprintf("%s=%v", k, v)
+			i++
+		}
+		traceContext := ";" + strings.Join(elems, ";")
+		h = h + traceContext
+	}
+
+	return h
+}
+
+// UnmarshalAmazonTraceContext parses the information provided in header and creates a
+// PropagationContext instance.
+func UnmarshalAmazonTraceContext(header string) (*PropagationContext, error) {
+	segments := strings.Split(header, ";")
+	// From https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-request-tracing.html
+	// If the X-Amzn-Trace-Id header is not present on an incoming request, the load balancer generates a header
+	// with a Root field and forwards the request. If the X-Amzn-Trace-Id header is present and has a Root field,
+	// the load balancer inserts a Self field, and forwards the request. If an application adds a header with a
+	// Root field and a custom field, the load balancer preserves both fields, inserts a Self field, and forwards
+	// the request. If the X-Amzn-Trace-Id header is present and has a Self field, the load balancer updates the
+	// value of the self field.
+	//
+	// Using the documentation above (that applies to amazon load balancers) we look for self as the parent id
+	// and root as the trace id. In the event that this context comes from a non-load balancer resource (e.g. a
+	// service instrumented with an X-Ray SDK) the parent segment ID will be included.
+	prop := &PropagationContext{}
+	prop.TraceContext = make(map[string]interface{})
+	for _, segment := range segments {
+		keyval := strings.SplitN(segment, "=", 2)
+		switch strings.ToLower(keyval[0]) {
+		case "self":
+			prop.ParentID = keyval[1]
+		case "root":
+			prop.TraceID = keyval[1]
+		case "parent":
+			prop.ParentID = keyval[1]
+		default:
+			prop.TraceContext[keyval[0]] = keyval[1]
+		}
+	}
+
+	// If no header is provided to an ALB or ELB, it will generate a header
+	// with a Root field and forwards the request. In this case it should be
+	// used as both the parent id and the trace id.
+	if prop.TraceID != "" && prop.ParentID == "" {
+		prop.ParentID = prop.TraceID
+	}
+
+	if prop.TraceID == "" && prop.ParentID != "" {
+		return nil, &PropagationError{"parent_id without trace_id", nil}
+	}
+
+	return prop, nil
```

Comment:
I opted to keep the behavior "return an empty propagationcontext object if the amazon header is missing or nil" but document it. I think there's a valid argument that "no data" is different than "invalid data". The honeycomb format specifies that there must be a version number, so no data is invalid data. If the Amazon header had the same requirement, I think it'd make sense to return an error. I don't feel strongly about it, but it made sense to me that way.

Reply:
I'm a little concerned at the variety of returned states available and the increased validation load this will put on the callers of these functions:
* when unmarshaling a honeycomb header, testing for `nil` error is enough; you can trust the prop
* when unmarshaling an amazon header, you must test for `nil` error and verify the prop isn't empty
* when unmarshaling a w3c header, there's no error but you have to test against `00000000` to see if it's empty

Each one is clearly documented, so maybe that's OK. Also, just as with function signatures, it emphasizes the reality that different serialization formats will provide different levels of insight into their success. 

Perhaps the easier answer than trying to assert consistency across the various unmarshaling implementations is to make it easier to test a `PropagationContext` to see if it's valid. Added that as a suggestion in the main propagation.go file.

Reply:
I'm letting this sit for a bit while I think about it. I think it's desirable to have consistent rules, and I'd ideally aim for "pass in a header that has a valid trace id and parent id or you'll get an error". Unfortunately, we've backed ourselves into a bit of a corner by allowing this with the Honeycomb header: 

https://github.com/honeycombio/beeline-go/blob/master/propagation/propagation_test.go#L51

So breaking that would be a breaking change. Not an API change, however, so we could probably get away with a major version bump and a note in the changelog about the change in behavior. I do note that we don't currently return an error in the W3C unmarshal function, so this would require changing the signature to:

`func UnmarshalW3CTraceContext(ctx context.Context, headers map[string]string) (context.Context, *PropagationContext, error)`

Which I'm not against. Just debating whether it's worth deprecating the old behavior.

Reply:
Okay, went with having a consistently applied restriction that you cannot create an invalid `PropagationContext` with an unmarshal function. Each of the `Unmarshal` functions now calls IsValid() and returns an error if the resulting `PropagationContext` contains neither a valid `TraceID` nor `ParentID`.